### ISSUE TITLE
Proposal for Fragment Results API

### ISSUE THREAD

**Issue by msasikanth** [simpledotorg/simple-android#2894]
## Problem

We are currently using our own solution for handling fragment to fragment communication, since there was no official way to do that, that fits our architecture last time we checked (#2219 ). While this approach works, it comes with couple of caveats.

### Screen results are not lifecycle aware

We dispatch the screen results after the fragment transactions are finished. Since the callback is not lifecycle aware, it's possible for the app to access views when they are not inflated or in backgroud which can cause crashes to the app.

### Bugs

We currently have a open bug [report](https://app.clubhouse.io/simpledotorg/story/3964/fix-race-condition-for-screen-results) for screen results, where sometimes the screen results are being casted to a different type and cause the app to crash. While it's easy to just check the type before casting, it shouldn't have happened in the first place.

When passing back the screen result, we try to make sure the target fragment is present. Sometimes that causes the app to [crash](https://simpledotorg.slack.com/archives/C01Q7S69AGL/p1627488837035700) when the fragment has not been attached to the activity or if the app has gone into background in the process.

### Extending a new type when creating Fragment

Since Fragment didn't had concept of screen results, we created a new type `ExpectsResult` to implement `onScreenResult` callback which receives the screen results. Accidentally missing extending this type can cause the app to crash.

## Goal

The goal is to avoid managing our own solution for manging communication between fragments, and to resolve the above mentioned issues.

## Approach

Android has created a [Fragment results API](https://developer.android.com/guide/fragments/communicate#fragment-result) which is available from version `1.3.0-alpha04`, which allows us to pass data between fragments. It solves some of the concerns we mentioned above, the results listener is lifecycle aware and only passes the results when the fragment is in started state.

The results listener API is baked into `FragmentManager`, which allows us to directly use it in fragments without having to extend new types.

### Listening to screen results

Fragment results API provides us with a lifecycle aware observer that let's us listen to results based on the request key
```kt
setFragmentResultListener(DatePickerResult) { result ->
  if (result is Succeeded) {
    val selectedDate = result.result as SelectedDate
    val event = ManualDateSelected(selectedDate = selectedDate.date, currentDate = LocalDate.now(userClock))
    hotEvents.onNext(event)
  }
}
```

For listening to multiple fragment results, we can simply add another listener with a different request key

```kt
setFragmentResultListener(DatePickerResult) { result ->
  if (result is Succeeded) {
    val selectedDate = result.result as SelectedDate
    val event = ManualDateSelected(selectedDate = selectedDate.date, currentDate = LocalDate.now(userClock))
    hotEvents.onNext(event)
  }
}

setFragmentResultListener(OverdueAppointmentRemoved) {
  router.pop()
}
```

### Sending the result

We can continue using the existing Router API for sending back the results, internal Router will send the result using `FragmentManager` which acts as a central store for all the fragment results.

```kt
router.popWithResult(Succeeded(SelectedDate(setDate)))
```

## Migration

The migration process for new fragment results API is pretty straightforward.

1. In `onViewCreated` of the fragment, we need to add a listener for the fragment results for the request key. If you have multiple request keys, you need to add multiple listeners. (We can use method references for result handling)

![image](https://user-images.githubusercontent.com/6140516/127428556-6e170bf8-1694-4243-8ad8-714a1b148a95.png)

2. Once you have added all the listeners for the request keys in the existing `onScreenResult`, we can stop extending `ExpectsResult` and remove `onScreenResult`

### Multiple listeners

Traditionally when we are handling screen results, we just have one single call back and that would return all the results. But with this approach we have to add multiple listeners to handle different requests.

**Comment by msasikanth** [comment]
Since we only have handful of screens that require screen results, we can finish the migration process in a single PR. #2896

**Comment by vinaysshenoy** [comment]
This approach looks good to me. Ideally, I'd like to abstract even the receiving of results via the `Router` in some way (similar to how we are abstracting the sending of results), but I think we can take a look at that after removing the existing mechanism for sending screen results since that will simplify a good bit of the `Router` code.

**Comment by msasikanth** [comment]
Okay, updated the `Router` to support that. The `dispatchScreenResult` function does get a bit complicated, but this would make use not to use the standard result pattern.

![image](https://user-images.githubusercontent.com/6140516/128295788-6e15b3d2-806d-43bd-9680-71f0eb762d0c.png)

Also, one of the consequences I have mentioned (multiple listeners), it's no longer a concern. We can change our extension to support multiple request keys and just iterate over them when setting listeners.

What do you think? I personally prefer the listener approach compared to this. But I am open to either approach.

@vinaysshenoy

**Comment by vinaysshenoy** [comment]
This actually looks good. Correct me if I am wrong though, but wouldn't the listening for results have to be done when the fragment is created though? How would this work with the lifecycle?

**Comment by msasikanth** [comment]
The listeners are lifecycle aware, so internally it will observe lifecycle and do necessary operations. So we will only get results when the state is started

![image](https://user-images.githubusercontent.com/6140516/128296560-a3c85e18-46c6-4bf8-8c6a-e083e0e678ef.png)

I have also updated the code to be a bit more concise and readable.

![image](https://user-images.githubusercontent.com/6140516/128296634-79d1caf2-ba1f-40d2-a779-dbe768174639.png)

**Comment by msasikanth** [comment]
As long as the fragment itself is initialised, we can get the lifecycles. Since we found the fragment from the fragment manager, it should be in the initialised state.

**Comment by Honey14** [comment]
Found easier to understand the extraction after the listeners implemented in `OnViewCreated`. 
Q: With `dispatchScreenResult`, we'd still have to implement the onScreenResult ToDo the body, right?

**Comment by msasikanth** [comment]
> Q: With `dispatchScreenResult`, we'd still have to implement the onScreenResult ToDo the body, right?

Yep, it will pass the request type and result and we can continue using the existing approach. So, only the internal implementation is changed.

**Comment by msasikanth** [comment]
The abstraction is causing a memory leak, I will see if we can resolve that leak, or else we will go with the listeners approach.

**Comment by msasikanth** [comment]
Was not able to resolve the memory leak, have to take a look at it at some other time. There seems to be some issue with the fragment lifecycle that we are observing for results in the `Router`. Going with the listener approach in the individual fragments.

cc: @vinaysshenoy @Honey14 @janhavisinghh

**Comment by msasikanth** [comment]
One small change is, I have updated the extensions to support multiple request keys.

![image](https://user-images.githubusercontent.com/6140516/128329662-b64ddedd-1cdf-42a9-a4d9-8f6bd844fcf7.png)

**Comment by msasikanth** [comment]
Closing this, since the change is merged in by #2948
